server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,492 @@
+        try {
+            MergeTask mergeTask = newMergeTask(mergeSource, merge, mergeTrigger);
+            return threadPoolMergeExecutorService.submitMergeTask(mergeTask);
+        } finally {
+            checkMergeTaskThrottling();
+        }
+    }
+
+    // package-private for tests
+    MergeTask newMergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, MergeTrigger mergeTrigger) {
+        // forced merges, as well as merges triggered when closing a shard, always run un-IO-throttled
+        boolean isAutoThrottle = mergeTrigger != MergeTrigger.CLOSING && merge.getStoreMergeInfo().mergeMaxNumSegments() == -1;
+        // IO throttling cannot be toggled for existing merge tasks, only new merge tasks pick up the updated IO throttling setting
+        return new MergeTask(
+            mergeSource,
+            merge,
+            isAutoThrottle && config.isAutoThrottle(),
+            "Lucene Merge Task #" + submittedMergeTaskCount.incrementAndGet() + " for shard " + shardId
+        );
+    }
+
+    private void checkMergeTaskThrottling() {
+        long submittedMergesCount = submittedMergeTaskCount.get();
+        long doneMergesCount = doneMergeTaskCount.get();
+        int executingMergesCount = currentlyRunningMergeTasks.size();
+        int configuredMaxMergeCount = config.getMaxMergeCount();
+        // both currently running and enqueued merge tasks are considered "active" for throttling purposes
+        int activeMerges = (int) (submittedMergesCount - doneMergesCount);
+        if (activeMerges > configuredMaxMergeCount && shouldThrottleIncomingMerges.get() == false) {
+            // maybe enable merge task throttling
+            synchronized (shouldThrottleIncomingMerges) {
+                if (shouldThrottleIncomingMerges.getAndSet(true) == false) {
+                    enableIndexingThrottling(executingMergesCount, activeMerges - executingMergesCount, configuredMaxMergeCount);
+                }
+            }
+        } else if (activeMerges <= configuredMaxMergeCount && shouldThrottleIncomingMerges.get()) {
+            // maybe disable merge task throttling
+            synchronized (shouldThrottleIncomingMerges) {
+                if (shouldThrottleIncomingMerges.getAndSet(false)) {
+                    disableIndexingThrottling(executingMergesCount, activeMerges - executingMergesCount, configuredMaxMergeCount);
+                }
+            }
+        }
+    }
+
+    // synchronized so that {@code #closed}, {@code #currentlyRunningMergeTasks} and {@code #backloggedMergeTasks} are modified atomically
+    private synchronized boolean runNowOrBacklog(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        if (closed) {
+            // Do not backlog or execute tasks when closing the merge scheduler, instead abort them.
+            mergeTask.abortOnGoingMerge();
+            throw new ElasticsearchException("merge task aborted because scheduler is shutting down");
```

Comment:
I am somewhat in doubt about throwing here or returning true. I get that if we return true we should not call `doMerge` in `abortOngoingMerge` when called from this callsite. I'd probably find it slightly more intuitive, since the exceptional termination here seems like a 3rd return value.

Reply:
I've refactored the `runNowOrBacklog` method to a different `schedule` method, that indeed has 3 return values: https://github.com/elastic/elasticsearch/pull/120869/commits/a217d12f78013fc45990adc962461899a1308817 .
Aborting and running merge tasks have to do some accounting and cleanup. Backlogging merge tasks doesn't need to do anything.

Reply:
Looks good, thanks.